org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/documentLink/LSPDocumentLinkPresentationReconcilingStrategy.java

```diff
@@ -70,18 +70,17 @@ private void underline() {
 		if (uri == null) {
 			return;
 		}
-		cancel();
-		final var params = new DocumentLinkParams(LSPEclipseUtils.toTextDocumentIdentifier(uri));
-		request = LanguageServiceAccessor
-				.getLanguageServers(document, capabilities -> capabilities.getDocumentLinkProvider() != null)
-				.thenAcceptAsync(languageServers -> CompletableFuture.allOf(languageServers.stream()
-						.map(languageServer -> languageServer.getTextDocumentService().documentLink(params))
-						.map(request -> request.thenAcceptAsync(links -> {
-							final Control control = viewer.getTextWidget();
-							if (control != null) {
-								control.getDisplay().asyncExec(() -> underline(links));
-							}
-						})).toArray(CompletableFuture[]::new)));
+		cancel();
+		final var params = new DocumentLinkParams(LSPEclipseUtils.toTextDocumentIdentifier(uri));
+		request = LanguageServers.forDocument(document)
+				.withFilter(capabilities -> capabilities.getDocumentLinkProvider() != null)
+				.collectAll(languageServer -> languageServer.getTextDocumentService().documentLink(params))
+				.thenAcceptAsync(links -> {

```

Comment:
Again, async not needed .... unless using the display as an executor

Reply:
.. which you can't do if control is null. Although whether that is something that should be guarded against is debatable. Could extract the display at the top of this method and then use it as the async executor.